mm2src/coins/hd_wallet/coin_ops.rs

```diff
@@ -0,0 +1,234 @@
+        derivation_path: DerivationPath,
+    ) -> HDCoinHDAddress<Self>;
+
+    /// Retrieves an HD address from the cache or derives it if it hasn't been derived yet.
+    fn derive_address_with_cache(
+        &self,
+        hd_account: &HDCoinHDAccount<Self>,
+        hd_addresses_cache: &mut HashMap<HDAddressId, HDCoinHDAddress<Self>>,
+        hd_address_id: HDAddressId,
+    ) -> AddressDerivingResult<HDCoinHDAddress<Self>> {
+        // Check if the given HD address has been derived already.
+        if let Some(hd_address) = hd_addresses_cache.get(&hd_address_id) {
+            return Ok(hd_address.clone());
+        }
+
+        let change_child = hd_address_id.chain.to_child_number();
+        let address_id_child = ChildNumber::from(hd_address_id.address_id);
+
+        let derived_pubkey = hd_account
+            .extended_pubkey()
+            .derive_child(change_child)?
+            .derive_child(address_id_child)?;
+
+        let mut derivation_path = hd_account.account_derivation_path();
+        derivation_path.push(change_child);
+        derivation_path.push(address_id_child);
+        drop_mutability!(derivation_path);
+        let hd_address = self.address_from_extended_pubkey(&derived_pubkey, derivation_path);
+
+        // Cache the derived `hd_address`.
+        hd_addresses_cache.insert(hd_address_id, hd_address.clone());
+        Ok(hd_address)
+    }
+
+    /// Derives a single HD address for a given account, chain, and address identifier.
+    async fn derive_address(
+        &self,
+        hd_account: &HDCoinHDAccount<Self>,
+        chain: Bip44Chain,
+        address_id: u32,
+    ) -> AddressDerivingResult<HDCoinHDAddress<Self>> {
+        self.derive_addresses(hd_account, std::iter::once(HDAddressId { chain, address_id }))
+            .await?
+            .into_iter()
+            .exactly_one()
+            // Unfortunately, we can't use [`MapToMmResult::map_to_mm`] due to unsatisfied trait bounds,
+            // and it's easier to use [`Result::map_err`] instead of adding more trait bounds to this method.
+            .map_err(|e| MmError::new(AddressDerivingError::Internal(e.to_string())))
+    }
+
+    /// Derives a set of HD addresses for a coin using the specified HD account and address identifiers.
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn derive_addresses<Ids>(
+        &self,
+        hd_account: &HDCoinHDAccount<Self>,
+        address_ids: Ids,
+    ) -> AddressDerivingResult<Vec<HDCoinHDAddress<Self>>>
+    where
+        Ids: Iterator<Item = HDAddressId> + Send,
+    {
+        let mut hd_addresses_cache_guard = hd_account.derived_addresses().lock().await;
+        let hd_addresses_cache = &mut *hd_addresses_cache_guard;
+        address_ids
+            .map(|hd_address_id| self.derive_address_with_cache(hd_account, hd_addresses_cache, hd_address_id))
+            .collect()
+    }
+
+    // Todo: combine both implementations once worker threads are supported in WASM
```

Comment:
Could you remind me please, as far as I remember we added worker support in wasm, right?
Do we need something else for this todo?

Reply:
Workers needs to be supported from the GUI side as well by adding a worker file for this and implementing any needed code if there is any.